subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1739 @@
+			}
+			/* wlk_addr will be the start research address in the next loop */
+			wlk_addr = rd_addr;
+		} else {
+			wlk_addr = fs->ate_wra;
+			break;
+		}
+	}
+
+	if (((!prev_found) || (wlk_ate.id != id)) || (wlk_ate.len == 0U) || (cnt_his < cnt)) {
+		return -ENOENT;
+	}
+
+	if (wlk_ate.len <= ZMS_DATA_IN_ATE_SIZE) {
+		/* data is stored in the ATE */
+		memcpy(data, &wlk_ate.data, MIN(len, wlk_ate.len));
+	} else {
+		rd_addr &= ADDR_SECT_MASK;
+		rd_addr += wlk_ate.offset;
+		rc = zms_flash_rd(fs, rd_addr, data, MIN(len, wlk_ate.len));
+		if (rc) {
+			goto err;
+		}
+#ifdef CONFIG_ZMS_DATA_CRC
+		computed_data_crc = crc32_ieee(data, wlk_ate.len);
+		if (computed_data_crc != wlk_ate.data_crc) {
+			LOG_ERR("Invalid data CRC: ATE_CRC=0x%08X, computed_data_crc=0x%08X",
+				wlk_ate.data_crc, computed_data_crc);
+			return -EIO;
+		}
+#endif
+	}
+
+	return MIN(len, wlk_ate.len);
+
+err:
+	return rc;
+}
+
+ssize_t zms_read(struct zms_fs *fs, uint32_t id, void *data, size_t len)
+{
+	int rc;
+
+	rc = zms_read_hist(fs, id, data, len, 0);
+	return rc;
+}
+
+ssize_t zms_calc_free_space(struct zms_fs *fs)
+{
+
+	int rc, previous_sector_num = ZMS_INVALID_SECTOR_NUM, prev_found = 0;
+	struct zms_ate step_ate, wlk_ate;
+	uint64_t step_addr, wlk_addr, step_prev_addr, wlk_prev_addr;
+	uint8_t current_cycle;
+	size_t ate_size, free_space;
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	free_space = 0;
+	for (uint32_t i = 1; i < fs->sector_count; i++) {
+		/*
+		 * There is always a closing ATE , an empty ATE, a GC done ATE
+		 * and a reserved ATE for deletion in each sector
+		 */
+		free_space += (fs->sector_size - 4 * ate_size);
+	}
+
+	step_addr = fs->ate_wra;
+
+	do {
+		step_prev_addr = step_addr;
+		rc = zms_prev_ate(fs, &step_addr, &step_ate);
+		if (rc) {
+			return rc;
+		}
+
+		/* When changing the sector let's get the new cycle counter */
+		rc = zms_get_cycle_on_sector_change(fs, step_prev_addr, previous_sector_num,
+						    &current_cycle);
+		if (rc) {
+			return rc;
+		}
+		previous_sector_num = SECTOR_NUM(step_prev_addr);
+
+		/* Invalid and deleted ATEs are free spaces.
+		 * Header ATEs are already retrieved from free space
+		 */
+		if (!zms_ate_valid_different_sector(fs, &step_ate, current_cycle) ||
+		    (step_ate.id == ZMS_HEAD_ID) || (step_ate.len == 0)) {
+			continue;
+		}
+
+		wlk_addr = step_addr;
+		/* Try to find if there is a previous valid ATE with same ID */
+		prev_found = zms_find_ate_with_id(fs, step_ate.id, wlk_addr, step_addr, &wlk_ate,
+						  &wlk_prev_addr);
+		if (prev_found < 0) {
+			return prev_found;
+		}
+
+		/* If no previous ATE is found than this is a valid ATE that cannot be GCed */
+		if (!prev_found || (wlk_prev_addr == step_prev_addr)) {
+			if (step_ate.len > ZMS_DATA_IN_ATE_SIZE) {
+				free_space -= zms_al_size(fs, step_ate.len);
+			}
+			free_space -= ate_size;
+		}
+	} while (step_addr != fs->ate_wra);
```

Comment:
Question : is the ATE at the address `fs->ate_wra` evaluated 2 times ? One at the entering of the loop and one more at the very end, when all the loop content is traversed upon reaching the `while` condition ?

Reply:
No the first time it enters the loop we are immediately increasing step_addr using the zms_prev_ate
So it is only evaluated once